File: R/cytoscape.r
#' Prepare files for building an enrichment map network visualization in Cytoscape
#'
#' This function writes four text files that are used to build an network using
#' Cytoscape and the EnrichmentMap app. The files are prefixed with \code{cytoscape_file_tag}. 
#'   The four files written are:
#'   \describe{
#'     \item{pathways.txt}{A list of significant terms and the
#'     associated p-value. Only terms with \code{adjusted_p_val <= significant} are
#'     written to this file}
#'     \item{subgroups.txt}{A matrix indicating whether the significant
#'     pathways are found to be significant when considering only one column (i.e., type of omics evidence) from
#'     \code{scores}. A 1 indicates that that term is significant using only that
#'     column to test for enrichment analysis}
#'     \item{pathways.gmt}{A shortened version of the supplied GMT
#'     file, containing only the terms in pathways.txt.}
#'     \item{legend.pdf}{A legend with colours matching contributions
#'     from columns in \code{scores}}
#'   }
#'
#' @param terms A data.table object with the columns 'term_id', 'term_name', 'adjusted_p_val'. 
#' @param gmt An abridged GMT object containing only the pathways that were
#' found to be significant in the ActivePathways analysis.
#' @param cytoscape_file_tag The user-defined file prefix and/or directory defining the location of the files.
#' @param col_significance A data.table object with a column 'term_id' and a column
#' for each type of omics evidence indicating whether a term was also found to be signficiant or not
#' when considering only the genes and p-values in the corresponding column of the \code{scores} matrix.
#' If term was not found, NA's are shown in columns, otherwise the relevant lists of genes are shown.
#' @param color_palette Color palette from RColorBrewer::brewer.pal to color each
#' column in the scores matrix. If NULL grDevices::rainbow is used by default.
#' @param custom_colors A character vector of custom colors for each column in the scores matrix.
#' @param color_integrated_only A character vector of length 1 specifying the color of the "combined" pathway contribution. 
#' @import ggplot2
#'
#' @return None

prepareCytoscape <- function(terms, 
                             gmt, 
                             cytoscape_file_tag, 
                             col_significance, color_palette = NULL, custom_colors = NULL, color_integrated_only = "#FFFFF0") {
  if (!is.null(col_significance)) {
    tests <- colnames(col_significance)[3:length(colnames(col_significance))]
    tests <- substr(tests, 7, 100)
    tests <- append(tests, "combined")
    
    rows <- 1:nrow(col_significance)
    
    evidence_columns = do.call(rbind, lapply(col_significance$evidence,
                                             function(x) 0+(tests %in% x)))
    colnames(evidence_columns) = tests
    
    col_significance = cbind(col_significance[,"term_id"], evidence_columns)
    
  
    if(is.null(color_palette) & is.null(custom_colors)) {
      col_colors <- grDevices::rainbow(length(tests))
    } else if (!is.null(custom_colors)){
        if (!is.null(names(custom_colors))){
          custom_colors <- custom_colors[order(match(names(custom_colors),tests))]
      }
        custom_colors <- append(custom_colors, color_integrated_only, after = match("combined",tests))
        col_colors <- custom_colors
    } else {
      col_colors <- RColorBrewer::brewer.pal(length(tests),color_palette)
    }
    col_colors <- replace(col_colors, match("combined",tests),color_integrated_only)
    if (!is.null(names(col_colors))){
      names(col_colors)[length(col_colors)] <- "combined"
    }
    
    instruct_str <- paste('piechart:',
                          ' attributelist="', 
                          paste(tests, collapse=','),
                          '" colorlist="', 
                          paste(col_colors, collapse=','), 
                          '" showlabels=FALSE', sep='')
    col_significance[, "instruct" := instruct_str]
    
    # Writing the Files
    utils::write.table(terms, 
                file=paste0(cytoscape_file_tag, "pathways.txt"), 
                row.names=FALSE, 
                sep="\t", 
                quote=FALSE)
    utils::write.table(col_significance, 
                file=paste0(cytoscape_file_tag, "subgroups.txt"), 
                row.names=FALSE, 
                sep="\t", 
                quote=FALSE)
    write.GMT(gmt, 
              paste0(cytoscape_file_tag, "pathways.gmt"))
    
    # Making a Legend
      dummy_plot = ggplot(data.frame("tests" = factor(tests, levels = tests),
                                     "value" = 1), aes(tests, fill = tests)) +
        geom_bar() +
        scale_fill_manual(name = "Contribution", values=col_colors)

      grDevices::pdf(file = NULL) # Suppressing Blank Display Device from ggplot_gtable
      dummy_table = ggplot_gtable(ggplot_build(dummy_plot))
      grDevices::dev.off()

      legend = dummy_table$grobs[[which(sapply(dummy_table$grobs, function(x) x$name) == "guide-box")]]
      
      # Estimating height & width
      legend_height = ifelse(length(tests) > 20, 
                             5.5, 
                             length(tests)*0.25+1)
      legend_width = ifelse(length(tests) > 20, 
                            ceiling(length(tests)/20)*(max(nchar(tests))*0.05+1), 
                            max(nchar(tests))*0.05+1)
      ggsave(legend,
             device = "pdf",
             filename = paste0(cytoscape_file_tag, "legend.pdf"), 
             height = legend_height, 
             width = legend_width, 
             scale = 1)
    
  } else {
    utils::write.table(terms, 
                file=paste0(cytoscape_file_tag, "pathways.txt"),
                row.names=FALSE, 
                sep="\t", 
                quote=FALSE)
    write.GMT(gmt, 
              paste0(cytoscape_file_tag, "pathways.gmt"))
  }
}

File: Python/cytoscape.py
import pandas as pd
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import matplotlib.cm
import csv

from gmt import write_gmt



def create_legend(color_dict, cytoscape_file_tag):
    # make a legend
    fig, ax = plt.subplots()

    #create a grid for positioning the rectangles
    grid = np.array([np.repeat(0.3, 2 * len(color_dict.keys())), np.flip(np.linspace(0.1,0.9,2*len(color_dict.keys())))]).T

    
    # rectangle dimension
    rec_dim = 0.1 / (len(color_dict.keys()) / 3)
    label_font_size = 12 / (len(color_dict.keys()) / 3)
        
    ax.annotate('Contribution', np.array(grid[0]) + np.array([0.15,0]), color='black', weight='bold', 
                fontsize=label_font_size, ha='center', va='center')

    # create rectangles
    for i, label in enumerate(list(color_dict.keys())):
        # define rectangle sizes
        rect = mpatches.Rectangle(grid[i + 1], rec_dim, rec_dim, color = color_dict[label])
        ax.add_patch(rect)
        
        # annotate
        ax.add_artist(rect)
        rx, ry = rect.get_xy()
        cx = rx + rect.get_width()/2
        cy = ry + rect.get_height()/2
        ax.annotate(label, (cx + 0.15, cy), color='black', weight='bold', 
                    fontsize=label_font_size, ha='center', va='center')



    plt.axis('off')
    outfile = cytoscape_file_tag + 'legend.pdf'
    plt.savefig(outfile)

def prepareCytoscape(terms, gmt, cytoscape_file_tag, col_significance = None, color_palette = None, custom_colors = None, color_integrated_only="#FFFFF0"):

    if type(col_significance) != None:
        # Obtain the name of each omics dataset and incorporate a 'combined' contribution
        tests = col_significance.columns.to_numpy(dtype='<U64')[2:]
        tests = list(map(lambda x: x[6:], tests))
        tests.append("combined")

        # Create a matrix of ones and zeros, where columns are omics datasets + 'combined' and rows are enriched pathways
        evidence_cols = pd.DataFrame()
        for i in range(len(tests)):
            evidence_cols.loc[:,tests[i]] = list(map(lambda x: int(tests[i] in x), col_significance['evidence']))
        #evidence_cols.loc[:,'term_id'] = col_significance['term_id'].to_numpy()
        evidence_cols.insert(loc=0, column='term_id', value=col_significance['term_id'].to_numpy())
        
        mapped_colors = []
        if color_palette == None and custom_colors == None:
            # Create default color palette
            cmap = mpl.cm.get_cmap('Set1', len(tests)-1)
            mapped_colors = list(map(lambda x: mpl.colors.rgb2hex(cmap(x)),range(len(tests)-1)))
            
        elif custom_colors != None:
            mapped_colors = custom_colors   
            
        # Create user-defined color pallette
        else:
            cmap = mpl.cm.get_cmap(color_palette, len(tests)-1)
            mapped_colors = list(map(lambda x: mpl.colors.rgb2hex(cmap(x)),range(len(tests)-1)))
            
        mapped_colors.append(color_integrated_only)

        # create a dictionary for colors
        color_dict = {}
        for i, color in enumerate(mapped_colors):
            color_dict[tests[i]] = color

        # cytoscape instructions to create pie charts
        instruct_str = f'''piechart: attributelist="{",".join(tests)}" colorlist="{",".join(mapped_colors)}" showlabels=FALSE'''
        evidence_cols.loc[:,'instruct'] = instruct_str
        
        # Writing the Files
        outfile = cytoscape_file_tag + "subgroups.txt"
        evidence_cols.to_csv(outfile, sep='\t', index=False, quoting=csv.QUOTE_NONE)

        create_legend(color_dict, cytoscape_file_tag)

    # if there is only one dataset, write results to files
    outfile = cytoscape_file_tag + "pathways.txt"
    terms.to_csv(outfile, sep='\t', index=False)

    write_gmt(gmt, (cytoscape_file_tag + "pathways.gmt"))
